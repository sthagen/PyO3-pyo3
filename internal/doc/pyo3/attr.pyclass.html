<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="A proc macro used to expose Rust structs and fieldless enums as Python objects."><meta name="keywords" content="rust, rustlang, rust-lang, pyclass"><title>pyclass in pyo3 - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../normalize.css"><link rel="stylesheet" type="text/css" href="../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../dark.css" disabled><link rel="stylesheet" type="text/css" href="../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../storage.js"></script><script defer src="sidebar-items.js"></script><script defer src="../main.js"></script><script defer src="../scrape-examples.js"></script><noscript><link rel="stylesheet" href="../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../favicon.svg"></head><body class="rustdoc attr"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../pyo3/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../pyo3/index.html"><div class="logo-container"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></div>
        </a><div class="sidebar-elems"><h2 class="location"><a href="index.html">In pyo3</a></h2></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../pyo3/index.html"><img class="rust-logo" src="../rust-logo.svg" alt="logo"></a><nav class="sub"><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><div id="help-button" title="help" tabindex="-1"><button type="button">?</button></div><div id="settings-menu" tabindex="-1">
                                <a href="../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../wheel.svg"></a></div>
                        </div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Attribute Macro <a href="index.html">pyo3</a>::<wbr><a class="attr" href="#">pyclass</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../src/pyo3_macros/lib.rs.html#83">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><div class="docblock item-decl"><pre class="rust attr"><code>#[pyclass]</code></pre></div><span class="item-info"><div class="stab portability">Available on <strong>crate feature <code>macros</code></strong> only.</div></span><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>A proc macro used to expose Rust structs and fieldless enums as Python objects.</p>
<p><code>#[pyclass]</code> can be used with the following parameters:</p>
<div><table><thead><tr><th style="text-align: left">Parameter</th><th style="text-align: left">Description</th></tr></thead><tbody>
<tr><td style="text-align: left"><span style="white-space: pre"><code>crate = &quot;some::path&quot;</code></span></td><td style="text-align: left">Path to import the <code>pyo3</code> crate, if it’s not accessible at <code>::pyo3</code>.</td></tr>
<tr><td style="text-align: left"><code>dict</code></td><td style="text-align: left">Gives instances of this class an empty <code>__dict__</code> to store custom attributes.</td></tr>
<tr><td style="text-align: left"><span style="white-space: pre"><code>extends = BaseType</code></span></td><td style="text-align: left">Use a custom baseclass. Defaults to <a href="https://docs.rs/pyo3/latest/pyo3/struct.PyAny.html"><code>PyAny</code></a></td></tr>
<tr><td style="text-align: left"><span style="white-space: pre"><code>freelist = N</code></span></td><td style="text-align: left">Implements a <a href="https://en.wikipedia.org/wiki/Free_list">free list</a> of size N. This can improve performance for types that are often created and deleted in quick succession. Profile your code to see whether <code>freelist</code> is right for you.</td></tr>
<tr><td style="text-align: left"><span style="white-space: pre"><code>frozen</code></span></td><td style="text-align: left">Declares that your pyclass is immutable. It removes the borrowchecker overhead when retrieving a shared reference to the Rust struct, but disables the ability to get a mutable reference.</td></tr>
<tr><td style="text-align: left"><code>mapping</code></td><td style="text-align: left">Inform PyO3 that this class is a <a href="https://pyo3.rs/latest/class/protocols.html#mapping--sequence-types"><code>Mapping</code></a>, and so leave its implementation of sequence C-API slots empty.</td></tr>
<tr><td style="text-align: left"><span style="white-space: pre"><code>module = &quot;module_name&quot;</code></span></td><td style="text-align: left">Python code will see the class as being defined in this module. Defaults to <code>builtins</code>.</td></tr>
<tr><td style="text-align: left"><span style="white-space: pre"><code>name = &quot;python_name&quot;</code></span></td><td style="text-align: left">Sets the name that Python sees this class as. Defaults to the name of the Rust struct.</td></tr>
<tr><td style="text-align: left"><span style="white-space: pre"><code>text_signature = &quot;(arg1, arg2, ...)&quot;</code></span></td><td style="text-align: left">Sets the text signature for the Python class’ <code>__new__</code> method.</td></tr>
<tr><td style="text-align: left"><code>subclass</code></td><td style="text-align: left">Allows other Python classes and <code>#[pyclass]</code> to inherit from this class. Enums cannot be subclassed.</td></tr>
<tr><td style="text-align: left"><code>unsendable</code></td><td style="text-align: left">Required if your struct is not <a href="https://doc.rust-lang.org/std/marker/trait.Send.html"><code>Send</code></a>. Rather than using <code>unsendable</code>, consider implementing your struct in a threadsafe way by e.g. substituting <a href="https://doc.rust-lang.org/std/rc/struct.Rc.html"><code>Rc</code></a> with <a href="https://doc.rust-lang.org/std/sync/struct.Arc.html"><code>Arc</code></a>. By using <code>unsendable</code>, your class will panic when accessed by another thread.</td></tr>
<tr><td style="text-align: left"><code>weakref</code></td><td style="text-align: left">Allows this class to be <a href="https://docs.python.org/3/library/weakref.html">weakly referenceable</a>.</td></tr>
</tbody></table>
</div>
<p>All of these parameters can either be passed directly on the <code>#[pyclass(...)]</code> annotation, or as one or
more accompanying <code>#[pyo3(...)]</code> annotations, e.g.:</p>

<div class='information'><div class='tooltip ignore'>ⓘ</div></div><div class="example-wrap"><pre class="rust rust-example-rendered ignore"><code><span class="comment">// Argument supplied directly to the `#[pyclass]` annotation.</span>
<span class="attribute">#[<span class="ident">pyclass</span>(<span class="ident">name</span> <span class="op">=</span> <span class="string">&quot;SomeName&quot;</span>, <span class="ident">subclass</span>)]</span>
<span class="kw">struct</span> <span class="ident">MyClass</span> { }

<span class="comment">// Argument supplied as a separate annotation.</span>
<span class="attribute">#[<span class="ident">pyclass</span>]</span>
<span class="attribute">#[<span class="ident">pyo3</span>(<span class="ident">name</span> <span class="op">=</span> <span class="string">&quot;SomeName&quot;</span>, <span class="ident">subclass</span>)]</span>
<span class="kw">struct</span> <span class="ident">MyClass</span> { }</code></pre></div>
<p>For more on creating Python classes,
see the <a href="https://pyo3.rs/latest/class.html">class section of the guide</a>.</p>
</div></details></section></div></main><div id="rustdoc-vars" data-root-path="../" data-current-crate="pyo3" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.64.0-nightly (3924dac7b 2022-07-29)" ></div>
</body></html>