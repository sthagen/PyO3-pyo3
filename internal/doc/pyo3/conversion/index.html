<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Defines conversions between Rust and Python types."><meta name="keywords" content="rust, rustlang, rust-lang, conversion"><title>pyo3::conversion - Rust</title><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Regular.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../FiraSans-Medium.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Regular.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceSerif4-Bold.ttf.woff2"><link rel="preload" as="font" type="font/woff2" crossorigin href="../../SourceCodePro-Semibold.ttf.woff2"><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled><link rel="stylesheet" type="text/css" href="../../dark.css" disabled><link rel="stylesheet" type="text/css" href="../../light.css" id="themeStyle"><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><script defer src="../../main.js"></script><script defer src="../../scrape-examples.js"></script>
    <noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../../favicon.svg"></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="mobile-topbar"><button class="sidebar-menu-toggle">&#9776;</button><a class="sidebar-logo" href="../../pyo3/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"></h2>
    </nav>
    <nav class="sidebar"><a class="sidebar-logo" href="../../pyo3/index.html"><div class="logo-container"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></div>
        </a><h2 class="location"><a href="#">Module conversion</a></h2><div class="sidebar-elems"><section><div class="block"><ul><li><a href="#traits">Traits</a></li></ul></div></section><div id="sidebar-vars" data-name="conversion" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><main><div class="width-limiter"><div class="sub-container"><a class="sub-logo-container" href="../../pyo3/index.html"><img class="rust-logo" src="../../rust-logo.svg" alt="logo"></a><nav class="sub"><div class="theme-picker hidden"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="22" height="22" alt="Pick another theme!" src="../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><form class="search-form"><div class="search-container"><span></span><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img width="22" height="22" alt="Change settings" src="../../wheel.svg"></a></div></form></nav></div><section id="main-content" class="content"><div class="main-heading">
    <h1 class="fqn"><span class="in-band">Module <a href="../index.html">pyo3</a>::<wbr><a class="mod" href="#">conversion</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span></h1><span class="out-of-band"><a class="srclink" href="../../src/pyo3/conversion.rs.html#3-584">source</a> · <a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Defines conversions between Rust and Python types.</p>
</div></details><h2 id="traits" class="small-section-header"><a href="#traits">Traits</a></h2>
<div class="item-table"><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.AsPyPointer.html" title="pyo3::conversion::AsPyPointer trait">AsPyPointer</a></div><div class="item-right docblock-short"><p>This trait represents that <strong>we can do zero-cost conversion from the object
to a FFI pointer</strong>.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.FromPyObject.html" title="pyo3::conversion::FromPyObject trait">FromPyObject</a></div><div class="item-right docblock-short"><p><code>FromPyObject</code> is implemented by various types that can be extracted from
a Python object reference.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.FromPyPointer.html" title="pyo3::conversion::FromPyPointer trait">FromPyPointer</a></div><div class="item-right docblock-short"><p>Raw level conversion between <code>*mut ffi::PyObject</code> and PyO3 types.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.IntoPy.html" title="pyo3::conversion::IntoPy trait">IntoPy</a></div><div class="item-right docblock-short"><p>Defines a conversion from a Rust type to a Python object.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.IntoPyPointer.html" title="pyo3::conversion::IntoPyPointer trait">IntoPyPointer</a></div><div class="item-right docblock-short"><p>This trait allows retrieving the underlying FFI pointer from Python objects.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.PyTryFrom.html" title="pyo3::conversion::PyTryFrom trait">PyTryFrom</a></div><div class="item-right docblock-short"><p>Trait implemented by Python object types that allow a checked downcast.
If <code>T</code> implements <code>PyTryFrom</code>, we can convert <code>&amp;PyAny</code> to <code>&amp;T</code>.</p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.PyTryInto.html" title="pyo3::conversion::PyTryInto trait">PyTryInto</a></div><div class="item-right docblock-short"><p>Trait implemented by Python object types that allow a checked downcast.
This trait is similar to <code>std::convert::TryInto</code></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.ToBorrowedObject.html" title="pyo3::conversion::ToBorrowedObject trait">ToBorrowedObject</a></div><div class="item-right docblock-short"><p>This trait has two implementations: The slow one is implemented for
all <a href="trait.ToPyObject.html" title="ToPyObject">ToPyObject</a> and creates a new object using <a href="trait.ToPyObject.html#tymethod.to_object" title="ToPyObject::to_object">ToPyObject::to_object</a>,
while the fast one is only implemented for AsPyPointer (we know
that every AsPyPointer is also ToPyObject) and uses <a href="trait.AsPyPointer.html#tymethod.as_ptr" title="AsPyPointer::as_ptr()">AsPyPointer::as_ptr()</a></p>
</div></div><div class="item-row"><div class="item-left module-item"><a class="trait" href="trait.ToPyObject.html" title="pyo3::conversion::ToPyObject trait">ToPyObject</a></div><div class="item-right docblock-short"><p>Conversion trait that allows various objects to be converted into <code>PyObject</code>.</p>
</div></div></div></section><section id="search" class="content hidden"></section></div></main><div id="rustdoc-vars" data-root-path="../../" data-current-crate="pyo3" data-themes="ayu,dark,light" data-resource-suffix="" data-rustdoc-version="1.62.0-nightly (60e50fc1c 2022-04-04)" ></div>
</body></html>