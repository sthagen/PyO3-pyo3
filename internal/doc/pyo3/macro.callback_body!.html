<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.callback_body.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.callback_body.html">macro.callback_body.html</a>...</p>
    <script>location.replace("macro.callback_body.html" + location.search + location.hash);</script>
</body>
</html>