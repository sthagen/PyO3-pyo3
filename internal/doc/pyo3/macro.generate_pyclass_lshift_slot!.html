<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=macro.generate_pyclass_lshift_slot.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="macro.generate_pyclass_lshift_slot.html">macro.generate_pyclass_lshift_slot.html</a>...</p>
    <script>location.replace("macro.generate_pyclass_lshift_slot.html" + location.search + location.hash);</script>
</body>
</html>