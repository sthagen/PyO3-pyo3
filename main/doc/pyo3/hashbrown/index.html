<!DOCTYPE html><html lang="en"><head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"><meta name="generator" content="rustdoc"><meta name="description" content="Conversions to and from hashbrown’s `HashMap` and `HashSet`."><meta name="keywords" content="rust, rustlang, rust-lang, hashbrown"><title>pyo3::hashbrown - Rust</title><link rel="stylesheet" type="text/css" href="../../normalize.css"><link rel="stylesheet" type="text/css" href="../../rustdoc.css" id="mainThemeStyle"><link rel="stylesheet" type="text/css" href="../../light.css"  id="themeStyle"><link rel="stylesheet" type="text/css" href="../../dark.css" disabled ><link rel="stylesheet" type="text/css" href="../../ayu.css" disabled ><script id="default-settings" ></script><script src="../../storage.js"></script><script src="../../crates.js"></script><noscript><link rel="stylesheet" href="../../noscript.css"></noscript><link rel="alternate icon" type="image/png" href="../../favicon-16x16.png"><link rel="alternate icon" type="image/png" href="../../favicon-32x32.png"><link rel="icon" type="image/svg+xml" href="../../favicon.svg"><style type="text/css">#crate-search{background-image:url("../../down-arrow.svg");}</style></head><body class="rustdoc mod"><!--[if lte IE 11]><div class="warning">This old browser is unsupported and will most likely display funky things.</div><![endif]--><nav class="sidebar"><div class="sidebar-menu" role="button">&#9776;</div><a href='../../pyo3/index.html'><div class='logo-container rust-logo'><img src='../../rust-logo.png' alt='logo'></div></a><h2 class="location">Module hashbrown</h2><div class="sidebar-elems"><div id="sidebar-vars" data-name="hashbrown" data-ty="mod" data-relpath="./"></div><script defer src="./sidebar-items.js"></script></div></nav><div class="theme-picker"><button id="theme-picker" aria-label="Pick another theme!" aria-haspopup="menu" title="themes"><img width="18" height="18" alt="Pick another theme!" src="../../brush.svg"></button><div id="theme-choices" role="menu"></div></div><nav class="sub"><form class="search-form"><div class="search-container"><div><select id="crate-search"><option value="All crates">All crates</option></select><input class="search-input" name="search" autocomplete="off" spellcheck="false" placeholder="Click or press ‘S’ to search, ‘?’ for more options…" type="search"></div><button type="button" id="help-button" title="help">?</button><a id="settings-menu" href="../../settings.html" title="settings"><img width="18" height="18" alt="Change settings" src="../../wheel.svg"></a></div></form></nav><section id="main" class="content"><h1 class="fqn"><span class="in-band">Module <a href="../index.html">pyo3</a>::<wbr><a class="mod" href="#">hashbrown</a><button id="copy-path" onclick="copy_path(this)" title="Copy item path to clipboard"><img src="../../clipboard.svg" width="19" height="18" alt="Copy item path"></button></span><span class="out-of-band"><span id="render-detail"><a id="toggle-all-docs" href="javascript:void(0)" title="collapse all docs">[<span class="inner">&#x2212;</span>]</a></span><a class="srclink" href="../../src/pyo3/conversions/hashbrown.rs.html#1-177" title="goto source code">[src]</a></span></h1><div class="item-info"><div class="stab portability">This is supported on <strong>crate feature <code>hashbrown</code></strong> only.</div></div><details class="rustdoc-toggle top-doc" open><summary class="hideme"><span>Expand description</span></summary><div class="docblock"><p>Conversions to and from <a href="https://docs.rs/hashbrown/">hashbrown</a>’s
<code>HashMap</code> and <code>HashSet</code>.</p>
<h2 id="setup" class="section-header"><a href="#setup">Setup</a></h2>
<p>To use this feature, add this to your <strong><code>Cargo.toml</code></strong>:</p>
<div class="example-wrap"><pre class="language-toml"><code>[dependencies]
hashbrown = &quot;*&quot;
pyo3 = { version = &quot;0.15.1&quot;, features = [&quot;hashbrown&quot;] }</code></pre></div>
<p>Note that you must use compatible versions of hashbrown and PyO3.
The required hashbrown version may vary based on the version of PyO3.</p>
</div></details></section><section id="search" class="content hidden"></section><div id="rustdoc-vars" data-root-path="../../" data-current-crate="pyo3" data-search-index-js="../../search-index.js" data-search-js="../../search.js"></div>
    <script src="../../main.js"></script>
</body></html>