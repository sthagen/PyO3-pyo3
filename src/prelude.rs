// Copyright (c) 2017-present PyO3 Project and Contributors

//! PyO3's prelude.
//!
//! The purpose of this module is to alleviate imports of many commonly used items of the PyO3 crate
//! by adding a glob import to the top of pyo3 heavy modules:
//!
//! ```
//! # #![allow(unused_imports)]
//! use pyo3::prelude::*;
//! ```

pub use crate::conversion::{
    FromPyObject, IntoPy, IntoPyPointer, PyTryFrom, PyTryInto, ToPyObject,
};
pub use crate::err::{PyErr, PyResult};
pub use crate::gil::GILGuard;
pub use crate::instance::{Py, PyObject};
pub use crate::pycell::{PyCell, PyRef, PyRefMut};
pub use crate::pyclass_init::PyClassInitializer;
pub use crate::python::Python;
pub use crate::types::{PyAny, PyModule};
pub use crate::wrap_pyfunction;

#[cfg(feature = "macros")]
pub use pyo3_macros::{pyclass, pyfunction, pymethods, pymodule, pyproto, FromPyObject};
