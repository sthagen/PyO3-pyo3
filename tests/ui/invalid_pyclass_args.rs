use pyo3::prelude::*;

#[pyclass(extend=pyo3::types::PyDict)]
struct TypoIntheKey {}

#[pyclass(extends = "PyDict")]
struct InvalidExtends {}

#[pyclass(name = m::MyClass)]
struct InvalidName {}

#[pyclass(name = "Custom Name")]
struct InvalidName2 {}

#[pyclass(name = CustomName)]
struct DeprecatedName {}

#[pyclass(rename_all = camelCase)]
struct InvalidRenamingRule {}

#[pyclass(rename_all = "Camel-Case")]
struct InvalidRenamingRule2 {}

#[pyclass(module = my_module)]
struct InvalidModule {}

#[pyclass(weakrev)]
struct InvalidArg {}

#[pyclass(mapping, sequence)]
struct CannotBeMappingAndSequence {}

fn main() {}
