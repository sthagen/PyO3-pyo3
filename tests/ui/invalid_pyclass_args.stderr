error: expected one of freelist/name/extends/module
 --> tests/ui/invalid_pyclass_args.rs:3:11
  |
3 | #[pyclass(extend=pyo3::types::PyDict)]
  |           ^^^^^^

error: expected type path (e.g., my_mod::BaseClass)
 --> tests/ui/invalid_pyclass_args.rs:6:21
  |
6 | #[pyclass(extends = "PyDict")]
  |                     ^^^^^^^^

error: expected type name (e.g. "Name")
 --> tests/ui/invalid_pyclass_args.rs:9:18
  |
9 | #[pyclass(name = m::MyClass)]
  |                  ^

error: expected a single identifier in double-quotes
  --> tests/ui/invalid_pyclass_args.rs:12:18
   |
12 | #[pyclass(name = "Custom Name")]
   |                  ^^^^^^^^^^^^^

error: since PyO3 0.13 a pyclass name should be in double-quotes, e.g. "CustomName"
  --> tests/ui/invalid_pyclass_args.rs:15:18
   |
15 | #[pyclass(name = CustomName)]
   |                  ^^^^^^^^^^

error: expected string literal (e.g., "my_mod")
  --> tests/ui/invalid_pyclass_args.rs:18:20
   |
18 | #[pyclass(module = my_module)]
   |                    ^^^^^^^^^

error: expected one of gc/weakref/subclass/dict/unsendable
  --> tests/ui/invalid_pyclass_args.rs:21:11
   |
21 | #[pyclass(weakrev)]
   |           ^^^^^^^
